[v1.13] node: Fix IP removal from ipset on node updates

qmonnet opened the pull request:
In CI, on upgrade/downgrade tests for IPsec, we have observed regular failures on some workflows for the node-to-node-encryption test. In this test, we look for some unencrypted packets from a client pod to a host on a distinct node; but sometimes, these packets do not show up.

Investigation showed that the packets do show up, but not with the expected IP address, thus failing to be caught by the tcpdump filter in use. In the failure state, instead of having the client pod IP as source address, packets have the origin host IP as source address. This is not supposed to happen: we use an iptables rule to exclude packets from masquerading in such case. Why then masquerading still occur? This is because the rule doesn't apply, the relevant host IP being missing in the ipset used by the rule.

Looking at the agent, the relevant node IP is missing from the ipset because on node updates, we sometimes remove it, when we should not. And the reason for that is that we're not comparing the right values to decide whether this removal should happen. Indeed, in the following:

    !slices.Contains(ipsAdded, address.IP.String())

The `address.IP.String()` represents a mere IP address, when the values added to ipsAdded are `netip.Prefix.String()`, containing the IP followed by a CIDR mask with the slash notation. For an example, `ipsAdded` can be `[172.19.0.2/32 fc00:c111::2/128]`, and `address.IP.String()` can be `172.19.0.2` (no CIDR mask): the search fails and we do remove the IP from the ipset, even if we just added it above in the function.

This commit fixes the comparison to ensure that we're looking for the relevant values in the slice, which in turns fixes the set of IPs contained in the ipset, and applies the masquerading exclusion to the relevant packets.

Related: https://github.com/cilium/cilium/issues/29351

qmonnet committed:
node: Fix IP removal from ipset on node updates

In CI, on upgrade/downgrade tests for IPsec, we have observed regular
failures on some workflows for the node-to-node-encryption test. In this
test, we look for some unencrypted packets from a client pod to a host
on a distinct node; but sometimes, these packets do not show up.

Investigation showed that the packets do show up, but not with the
expected IP address, thus failing to be caught by the tcpdump filter in
use. In the failure state, instead of having the client pod IP as source
address, packets have the origin host IP as source address. This is not
supposed to happen: we use an iptables rule to exclude packets from
masquerading in such case. Why then masquerading still occur? This is
because the rule doesn't apply, the relevant host IP being missing in
the ipset used by the rule.

Looking at the agent, the relevant node IP is missing from the ipset
because on node updates, we sometimes remove it, when we should not. And
the reason for that is that we're not comparing the right values to
decide whether this removal should happen. Indeed, in the following:

    !slices.Contains(ipsAdded, address.IP.String())

The address.IP.String() represents a mere IP address, when the values
added to ipsAdded are netip.Prefix.String(), containing the IP followed
by a CIDR mask with the slash notation. For an example, ipsAdded can be
[172.19.0.2/32 fc00:c111::2/128], and address.IP.String() can be
172.19.0.2 (no CIDR mask): the search fails and we do remove the IP from
the ipset, even if we just added it above in the function.

This commit fixes the comparison to ensure that we're looking for the
relevant values in the slice, which in turns fixes the set of IPs
contained in the ipset, and applies the masquerading exclusion to the
relevant packets.

Fixes: 454df0c065bd ("ipcache/md: Convert metadata key to netip.Prefix")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
pkg/node/manager: Add test to validate ipset entries

Following a bug in the way we update the list of ipset entries in the
node manager, add a test to ensure that entries are added or removed as
we expect on node updates.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

brb reviewed (approved):
Nice! :rocket:

gandro reviewed (approved):
The pure nodemanager change looks good to me. Do we need to forward-port this to `main`?

qmonnet committed:
.travis: Install ipset on runner

We want to check entries in ipsets on the machine as part of one of the
runtime test. This requires the ipset binary to be available on the
runner.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>
